python/lsst/daf/butler/queries/tree/_query_tree.py

```diff
@@ -0,0 +1,314 @@
+from ._base import InvalidQueryError, QueryTreeBase
+from ._column_set import ColumnSet
+from ._predicate import Predicate
+
+DataCoordinateUploadKey: TypeAlias = uuid.UUID
+
+MaterializationKey: TypeAlias = uuid.UUID
+
+
+def make_identity_query_tree(universe: DimensionUniverse) -> QueryTree:
+    """Make an initial query tree with empty dimensions and a single logical
+    row.
+
+    This method should be used by `Butler._query` to construct the initial
+    query tree.  This tree is a useful initial state because it is the
+    identity for joins, in that joining any other query tree to the identity
+    yields that query tree.
+
+    Parameters
+    ----------
+    universe : `..DimensionUniverse`
+        Definitions for all dimensions.
+
+    Returns
+    -------
+    tree : `QueryTree`
+        A tree with empty dimensions.
+    """
+    return QueryTree(dimensions=universe.empty.as_group())
+
+
+@final
+class DatasetSearch(QueryTreeBase):
+    """Information about a dataset search joined into a query tree.
+
+    The dataset type name is the key of the dictionary (in `QueryTree`) where
+    this type is used as a value.
+    """
+
+    collections: tuple[str, ...]
+    """The collections to search.
+
+    Order matters if this dataset type is later referenced by a `FindFirst`
+    operation.  Collection wildcards are always resolved before being included
+    in a dataset search.
+    """
+
+    dimensions: DimensionGroup
+    """The dimensions of the dataset type.
+
+    This must match the dimensions of the dataset type as already defined in
+    the butler database, but this cannot generally be verified when a relation
+    tree is validated (since it requires a database query) and hence must be
+    checked later.
+    """
+
+    storage_class_name: str | None
+    """Name of the storage class to use when returning `DatasetRef` results.
+
+    May be `None` if the dataset is only used as a constraint or to return
+    columns that do not include a full dataset type.
+    """
+
+
+@final
+class QueryTree(QueryTreeBase):
+    """A declarative, serializable description of the row constraints and joins
+    in a butler query.
+
+    Notes
+    -----
+    A `QueryTree` is the struct that represents the serializable form of a
+    `Query` object, or one piece (with `ResultSpec` the other) of the
+    serializable form of a query results object.
+
+    This class's attributes describe the columns that are "available" to be
+    returned or used in ``where`` or ``order_by`` expressions, but it does not
+    carry information about the columns that are actually included in result
+    rows, or what kind of butler primitive (e.g. `DataCoordinate` or
+    `DatasetRef`) those rows might be transformed into.
+    """
+
+    dimensions: DimensionGroup
+    """The dimensions whose keys are joined into the query.
+    """
+
+    datasets: Mapping[str, DatasetSearch] = pydantic.Field(default_factory=dict)
```

Comment:
I'm using `Mapping` just because there's no (built-in, Pydantic-friendly) `frozendict` and I want mypy to prevent in-place modifications.  I'm open to other ideas.

Reply:
That's reasonable -- I missed the frozendicts, and I forgot that `Mapping` and `MutableMapping` are two separate things.